File: daemon/internal/connectors/slack/readiness_test.go
package slack

import (
	"testing"
	"time"

	baseconnectors "github.com/dopejs/dope-agent/daemon/internal/connectors"
)

func TestEvaluateHostedSetupReadyWithValidWorkspaceAndRoutePolicy(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 5, 8, 10, 0, 0, 0, time.UTC)
	setup := EvaluateHostedSetup(HostedSetupInput{
		TenantID:          "ten_slack",
		ConnectorID:       "slack-main",
		DisplayName:       "Slack Main",
		OAuthState:        OAuthGrantValid,
		ProviderAvailable: true,
		NetworkAvailable:  true,
		ValidatedAt:       now,
		WorkspaceBinding: WorkspaceBinding{
			WorkspaceID:        "workspace_redacted",
			InstallationID:     "installation_redacted",
			OAuthGrantState:    "valid",
			RequiredScopeState: "valid",
		},
		RoutePolicy: RoutePolicy{
			ValidationState: RoutePolicyValid,
			SelectedChannels: []ConversationRoute{{
				ConversationID:       "channel_redacted",
				ConversationType:     ConversationChannel,
				SelectedChannelState: SelectedChannelSelected,
				ValidationState:      RoutePolicyValid,
			}},
		},
	})
	if setup.TerminalState != TerminalReady || setup.Status != baseconnectors.LifecycleStateHealthy || !setup.DeliveryEligible {
		t.Fatalf("expected ready hosted setup, got %+v", setup)
	}
	if setup.WorkspaceBinding.WorkspaceBindingID != "slack_workspace_slack-main" || setup.RoutePolicyState != RoutePolicyStateValid {
		t.Fatalf("expected normalized workspace binding and route policy, got %+v", setup)
	}
}

func TestEvaluateHostedSetupMapsOAuthTerminalStates(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 5, 8, 10, 0, 0, 0, time.UTC)
	cases := []struct {
		name       string
		input      HostedSetupInput
		wantState  TerminalState
		wantOAuth  OAuthState
		wantReason string
	}{
		{name: "missing grant", input: HostedSetupInput{OAuthState: OAuthGrantMissing}, wantState: TerminalActionRequired, wantOAuth: OAuthGrantMissing, wantReason: string(baseconnectors.DiagnosticAuthMissing)},
		{name: "revoked", input: HostedSetupInput{OAuthState: OAuthRevoked}, wantState: TerminalActionRequired, wantOAuth: OAuthRevoked, wantReason: string(baseconnectors.DiagnosticAuthMissing)},
		{name: "scope missing", input: HostedSetupInput{OAuthState: OAuthScopeMissing}, wantState: TerminalActionRequired, wantOAuth: OAuthScopeMissing, wantReason: string(baseconnectors.DiagnosticPermissionMissing)},
		{name: "approval required", input: HostedSetupInput{OAuthState: OAuthApprovalRequired}, wantState: TerminalActionRequired, wantOAuth: OAuthApprovalRequired, wantReason: string(baseconnectors.DiagnosticPermissionMissing)},
		{name: "cancelled", input: HostedSetupInput{OAuthState: OAuthStarted, Cancelled: true}, wantState: TerminalCancelled, wantOAuth: OAuthStarted, wantReason: "user_cancelled"},
		{name: "provider unavailable", input: HostedSetupInput{OAuthState: OAuthGrantValid, ProviderAvailable: false, NetworkAvailable: true}, wantState: TerminalUnavailable, wantOAuth: OAuthGrantValid, wantReason: string(baseconnectors.DiagnosticProviderUnavailable)},
		{name: "network failed", input: HostedSetupInput{OAuthState: OAuthGrantValid, ProviderAvailable: true, NetworkAvailable: false}, wantState: TerminalUnavailable, wantOAuth: OAuthGrantValid, wantReason: string(baseconnectors.DiagnosticNetworkFailed)},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			input := tc.input
			input.TenantID = "ten_slack"
			input.ConnectorID = "slack-main"
			input.DisplayName = "Slack Main"
			input.ValidatedAt = now
			setup := EvaluateHostedSetup(input)
			if setup.TerminalState != tc.wantState || setup.OAuthState != tc.wantOAuth || setup.ReasonCode != tc.wantReason || setup.DeliveryEligible {
				t.Fatalf("setup=%+v, want state=%s oauth=%s reason=%s and ineligible", setup, tc.wantState, tc.wantOAuth, tc.wantReason)
			}
		})
	}
}

func TestEvaluateHostedSetupRequiresValidatedRoutePolicy(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 5, 8, 10, 0, 0, 0, time.UTC)
	base := HostedSetupInput{
		TenantID:          "ten_slack",
		ConnectorID:       "slack-main",
		DisplayName:       "Slack Main",
		OAuthState:        OAuthGrantValid,
		ProviderAvailable: true,
		NetworkAvailable:  true,
		ValidatedAt:       now,
		WorkspaceBinding: WorkspaceBinding{
			WorkspaceID:        "workspace_redacted",
			InstallationID:     "installation_redacted",
			OAuthGrantState:    "valid",
			RequiredScopeState: "valid",
		},
	}
	setup := EvaluateHostedSetup(base)
	if setup.TerminalState != TerminalActionRequired || setup.RoutePolicyState != RoutePolicyStateNone || setup.DeliveryEligible {
		t.Fatalf("valid OAuth without selected route must stay action-required, got %+v", setup)
	}
	base.RoutePolicy = RoutePolicy{
		ValidationState: RoutePolicyBlocked,
		AllowedDMUsers:  []string{"user_hash_1"},
	}
	setup = EvaluateHostedSetup(base)
	if setup.TerminalState != TerminalActionRequired || setup.DeliveryEligible {
		t.Fatalf("blocked route policy with DM allowment must not be ready, got %+v", setup)
	}
}

func TestEvaluateHostedSetupWorkspaceBindingCardinalityAndTimeout(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 5, 8, 10, 6, 0, 0, time.UTC)
	setup := EvaluateHostedSetup(HostedSetupInput{
		TenantID:            "ten_slack",
		ConnectorID:         "slack-main",
		DisplayName:         "Slack Main",
		OAuthState:          OAuthGrantValid,
		ProviderAvailable:   true,
		NetworkAvailable:    true,
		ExpectedWorkspaceID: "workspace_expected",
		StartedAt:           now.Add(-6 * time.Minute),
		ValidatedAt:         now,
		WorkspaceBinding: WorkspaceBinding{
			WorkspaceID:        "workspace_other",
			InstallationID:     "installation_redacted",
			OAuthGrantState:    "valid",
			RequiredScopeState: "valid",
		},
		RoutePolicy: RoutePolicy{
			ValidationState: RoutePolicyValid,
			AllowedDMUsers:  []string{"user_hash_1"},
		},
	})
	if setup.TerminalState != TerminalUnavailable || setup.ReasonCode != "setup_timeout" {
		t.Fatalf("expected timeout to fail with actionable terminal state before readiness, got %+v", setup)
	}

	first := EvaluateHostedSetup(HostedSetupInput{TenantID: "ten_slack", ConnectorID: "slack-east", ValidatedAt: now})
	second := EvaluateHostedSetup(HostedSetupInput{TenantID: "ten_slack", ConnectorID: "slack-west", ValidatedAt: now})
	if first.WorkspaceBindingID == second.WorkspaceBindingID {
		t.Fatalf("multiple connectors for one tenant need distinct workspace bindings: %s", first.WorkspaceBindingID)
	}

	mismatch := EvaluateHostedSetup(HostedSetupInput{
		TenantID:            "ten_slack",
		ConnectorID:         "slack-main",
		OAuthState:          OAuthGrantValid,
		ProviderAvailable:   true,
		NetworkAvailable:    true,
		ExpectedWorkspaceID: "workspace_expected",
		ValidatedAt:         now,
		WorkspaceBinding: WorkspaceBinding{
			WorkspaceID:        "workspace_other",
			InstallationID:     "installation_redacted",
			OAuthGrantState:    "valid",
			RequiredScopeState: "valid",
		},
		RoutePolicy: RoutePolicy{
			ValidationState: RoutePolicyValid,
			AllowedDMUsers:  []string{"user_hash_1"},
		},
	})
	if mismatch.TerminalState != TerminalActionRequired || mismatch.ReasonCode != "workspace_mismatch" || mismatch.DeliveryEligible {
		t.Fatalf("expected workspace mismatch to fail closed, got %+v", mismatch)
	}
}

File: crates/channels/slack/tests/readiness_test.rs
//! Behavioral tests for the hosted-setup readiness state machine (port of
//! readiness_test.go).

use chrono::{Duration, TimeZone, Utc};
use kura_connectors::{DiagnosticReasonCode, LifecycleState};
use kura_slack::destinations::{
    ConversationRoute, ConversationType, RoutePolicy, RouteValidationState, SelectedChannelState,
};
use kura_slack::readiness::{
    HostedSetupInput, OAuthState, RoutePolicyState, TerminalState, WorkspaceBinding,
    evaluate_hosted_setup,
};

fn ts(y: i32, mo: u32, d: u32, h: u32, mi: u32) -> chrono::DateTime<Utc> {
    Utc.with_ymd_and_hms(y, mo, d, h, mi, 0)
        .single()
        .expect("valid timestamp")
}

fn ready_binding() -> WorkspaceBinding {
    WorkspaceBinding {
        workspace_id: "workspace_redacted".to_string(),
        installation_id: "installation_redacted".to_string(),
        oauth_grant_state: "valid".to_string(),
        required_scope_state: "valid".to_string(),
        ..WorkspaceBinding::default()
    }
}

fn ready_route_policy() -> RoutePolicy {
    RoutePolicy {
        validation_state: RouteValidationState::Valid,
        selected_channels: vec![ConversationRoute {
            conversation_id: "channel_redacted".to_string(),
            conversation_type: ConversationType::Channel,
            selected_channel_state: SelectedChannelState::Selected,
            validation_state: RouteValidationState::Valid,
            ..ConversationRoute::default()
        }],
        ..RoutePolicy::default()
    }
}

#[test]
fn evaluate_hosted_setup_ready_with_valid_workspace_and_route_policy() {
    let now = ts(2026, 5, 8, 10, 0);
    let setup = evaluate_hosted_setup(HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        display_name: "Slack Main".to_string(),
        oauth_state: OAuthState::GrantValid,
        provider_available: true,
        network_available: true,
        validated_at: now,
        workspace_binding: ready_binding(),
        route_policy: ready_route_policy(),
        ..HostedSetupInput::default()
    });
    assert_eq!(
        setup.terminal_state,
        TerminalState::Ready,
        "expected ready hosted setup"
    );
    assert_eq!(setup.status, LifecycleState::Healthy);
    assert!(setup.delivery_eligible);
    assert_eq!(
        setup
            .workspace_binding
            .as_ref()
            .map(|b| b.workspace_binding_id.as_str()),
        Some("slack_workspace_slack-main"),
        "expected normalized workspace binding"
    );
    assert_eq!(setup.route_policy_state, RoutePolicyState::Valid);
}

#[test]
fn evaluate_hosted_setup_maps_oauth_terminal_states() {
    let now = ts(2026, 5, 8, 10, 0);
    let cases: Vec<(&str, HostedSetupInput, TerminalState, OAuthState, &str)> = vec![
        (
            "missing grant",
            HostedSetupInput {
                oauth_state: OAuthState::GrantMissing,
                ..HostedSetupInput::default()
            },
            TerminalState::ActionRequired,
            OAuthState::GrantMissing,
            DiagnosticReasonCode::AuthMissing.as_str(),
        ),
        (
            "revoked",
            HostedSetupInput {
                oauth_state: OAuthState::Revoked,
                ..HostedSetupInput::default()
            },
            TerminalState::ActionRequired,
            OAuthState::Revoked,
            DiagnosticReasonCode::AuthMissing.as_str(),
        ),
        (
            "scope missing",
            HostedSetupInput {
                oauth_state: OAuthState::ScopeMissing,
                ..HostedSetupInput::default()
            },
            TerminalState::ActionRequired,
            OAuthState::ScopeMissing,
            DiagnosticReasonCode::PermissionMissing.as_str(),
        ),
        (
            "approval required",
            HostedSetupInput {
                oauth_state: OAuthState::ApprovalRequired,
                ..HostedSetupInput::default()
            },
            TerminalState::ActionRequired,
            OAuthState::ApprovalRequired,
            DiagnosticReasonCode::PermissionMissing.as_str(),
        ),
        (
            "cancelled",
            HostedSetupInput {
                oauth_state: OAuthState::Started,
                cancelled: true,
                ..HostedSetupInput::default()
            },
            TerminalState::Cancelled,
            OAuthState::Started,
            "user_cancelled",
        ),
        (
            "provider unavailable",
            HostedSetupInput {
                oauth_state: OAuthState::GrantValid,
                provider_available: false,
                network_available: true,
                ..HostedSetupInput::default()
            },
            TerminalState::Unavailable,
            OAuthState::GrantValid,
            DiagnosticReasonCode::ProviderUnavailable.as_str(),
        ),
        (
            "network failed",
            HostedSetupInput {
                oauth_state: OAuthState::GrantValid,
                provider_available: true,
                network_available: false,
                ..HostedSetupInput::default()
            },
            TerminalState::Unavailable,
            OAuthState::GrantValid,
            DiagnosticReasonCode::NetworkFailed.as_str(),
        ),
    ];
    for (name, mut input, want_state, want_oauth, want_reason) in cases {
        input.tenant_id = "ten_slack".to_string();
        input.connector_id = "slack-main".to_string();
        input.display_name = "Slack Main".to_string();
        input.validated_at = now;
        let setup = evaluate_hosted_setup(input);
        assert_eq!(setup.terminal_state, want_state, "{name}: terminal state");
        assert_eq!(setup.oauth_state, want_oauth, "{name}: oauth state");
        assert_eq!(setup.reason_code, want_reason, "{name}: reason");
        assert!(
            !setup.delivery_eligible,
            "{name}: must be delivery-ineligible"
        );
    }
}

#[test]
fn evaluate_hosted_setup_requires_validated_route_policy() {
    let now = ts(2026, 5, 8, 10, 0);
    let base = HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        display_name: "Slack Main".to_string(),
        oauth_state: OAuthState::GrantValid,
        provider_available: true,
        network_available: true,
        validated_at: now,
        workspace_binding: ready_binding(),
        ..HostedSetupInput::default()
    };
    let setup = evaluate_hosted_setup(base.clone());
    assert_eq!(setup.terminal_state, TerminalState::ActionRequired);
    assert_eq!(setup.route_policy_state, RoutePolicyState::None);
    assert!(
        !setup.delivery_eligible,
        "valid OAuth without selected route must stay action-required"
    );

    let blocked = HostedSetupInput {
        route_policy: RoutePolicy {
            validation_state: RouteValidationState::Blocked,
            allowed_dm_users: vec!["user_hash_1".to_string()],
            ..RoutePolicy::default()
        },
        ..base
    };
    let setup = evaluate_hosted_setup(blocked);
    assert_eq!(setup.terminal_state, TerminalState::ActionRequired);
    assert!(
        !setup.delivery_eligible,
        "blocked route policy with DM allowment must not be ready"
    );
}

#[test]
fn evaluate_hosted_setup_workspace_binding_cardinality_and_timeout() {
    let now = ts(2026, 5, 8, 10, 6);
    let timeout_setup = evaluate_hosted_setup(HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        display_name: "Slack Main".to_string(),
        oauth_state: OAuthState::GrantValid,
        provider_available: true,
        network_available: true,
        expected_workspace_id: "workspace_expected".to_string(),
        started_at: now - Duration::minutes(6),
        validated_at: now,
        workspace_binding: WorkspaceBinding {
            workspace_id: "workspace_other".to_string(),
            installation_id: "installation_redacted".to_string(),
            oauth_grant_state: "valid".to_string(),
            required_scope_state: "valid".to_string(),
            ..WorkspaceBinding::default()
        },
        route_policy: RoutePolicy {
            validation_state: RouteValidationState::Valid,
            allowed_dm_users: vec!["user_hash_1".to_string()],
            ..RoutePolicy::default()
        },
        ..HostedSetupInput::default()
    });
    assert_eq!(timeout_setup.terminal_state, TerminalState::Unavailable);
    assert_eq!(timeout_setup.reason_code, "setup_timeout");

    let first = evaluate_hosted_setup(HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-east".to_string(),
        validated_at: now,
        ..HostedSetupInput::default()
    });
    let second = evaluate_hosted_setup(HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-west".to_string(),
        validated_at: now,
        ..HostedSetupInput::default()
    });
    assert_ne!(
        first.workspace_binding_id, second.workspace_binding_id,
        "multiple connectors for one tenant need distinct workspace bindings"
    );

    let mismatch = evaluate_hosted_setup(HostedSetupInput {
        tenant_id: "ten_slack".to_string(),
        connector_id: "slack-main".to_string(),
        oauth_state: OAuthState::GrantValid,
        provider_available: true,
        network_available: true,
        expected_workspace_id: "workspace_expected".to_string(),
        validated_at: now,
        workspace_binding: WorkspaceBinding {
            workspace_id: "workspace_other".to_string(),
            installation_id: "installation_redacted".to_string(),
            oauth_grant_state: "valid".to_string(),
            required_scope_state: "valid".to_string(),
            ..WorkspaceBinding::default()
        },
        route_policy: RoutePolicy {
            validation_state: RouteValidationState::Valid,
            allowed_dm_users: vec!["user_hash_1".to_string()],
            ..RoutePolicy::default()
        },
        ..HostedSetupInput::default()
    });
    assert_eq!(mismatch.terminal_state, TerminalState::ActionRequired);
    assert_eq!(mismatch.reason_code, "workspace_mismatch");
    assert!(
        !mismatch.delivery_eligible,
        "expected workspace mismatch to fail closed"
    );
}
